sycl/doc/design/CommandGraph.md

```diff
@@ -0,0 +1,204 @@
+CG object of a specific type. During normal operation, `handler::finalize()`
+then passes the CG object to the scheduler, and a `sycl::event` object
+representing the command-group is returned.
+
+However during graph construction, inside `hander::finalize()` the CG object is
+not submitted for execution as normal, but stored in the graph as a new node
+instead.
+
+When a user adds a node to a graph using the explicit
+`command_graph<modifiable>::add()` API passing a CGF, in our graph runtime
+implementation a `sycl::handler` object is constructed with a graph parameter
+telling it to not submit the CG object to the scheduler on finalization.
+This handler finalizes the CGF, and after finalization the CG object from the
+handler is moved to the node.
+
+For creating a node in the graph using queue recording mode. When the
+`sycl::handler` from a queue submission is finalized, if the queue the
+handler was created from is in the recording mode, then the handler knows
+not to submit the CG object to the scheduler. Instead, the CG object is
+added to the graph associated with the queue as a new node.
+
+Edges are stored in each node as lists of predecessor and successor nodes.
+
+## Execution Order
+
+The current way graph nodes are linearized into execution order is using a
+reversed depth-first sorting algorithm. Alternative algorithms, such as
+breadth-first, are possible and may give better performance on certain
+workloads/hardware. In the future there might be options for allowing the
+user to control this implementation detail.
+
+## Scheduler Integration
+
+When there are no requirements from accessors in a command-graph submission,
+the scheduler is bypassed and the underlying UR command-buffer is directly
+enqueued to a UR queue. If there are accessor requirements, the UR
+command-buffer for the executable graph needs to be enqueued by the scheduler.
+
+When individual graph nodes have requirements from SYCL accessors, the
+underlying `sycl::detail::CG` object stored in the node is copied and passed to
+the scheduler for adding to the UR command-buffer, otherwise the node can
+be appended directly as a command in the UR command-buffer. This is in-keeping
+with the existing behaviour of the handler with normal queue submissions.
+
+## Memory handling: Buffer and Accessor
+
+There is no extra support for Graph-specific USM allocations in the current
+proposal. Memory operations will be supported subsequently by the current
+implementation starting with `memcpy`.
+
+Buffers and accessors are supported in a command-graph. The following restrictions
+are required to adapt buffers and their lifetime to a lazy work execution model:
+
+- The lifetime of a buffer with host data will be extended by copying the underlying
+data.
+- Host accessors on buffers that are currently used by a command-graph are prohibited.
+- Copy-back behavior on destruction of a buffer is prohibited.
```

Comment:
Prohibited means a user needs to set `buffer::set_final_data(nullptr)` or `buffer::set_write_back(false)` on a buffer used in the graph. But yes, this is really just repeating spec information rather than giving implementation insight, so i'll replace this bulleted list with a reference to the relevant spec section.